include/mxnet/ndarray.h

```diff
@@ -74,6 +74,7 @@ enum NDArrayFormatErr {
   kRSPIdxErr,     // indices error for row sparse
 };
 
+class MKLDNNMemory;
```

Comment:
forward declarations are going out of style.  Is there a reasonable way around this, or does it get messy without this?

Reply:
i don't think it's a good idea to include the header file that defines MKLDNNMemory in ndarray.h
it's better to declare the class in this file. what is the reason of not using forward declaration?

Reply:
one reason is that they cause annoying compile errors when used in pointer classes when code the compiler decides it needs the type in order to generate the destructor code, for instance or during template instantiation of something that uses it directly or indirectly. i’m not going to block the PR over it and if you feel
strongly that you want to use it then fine, but it’s not done much in the code base and that’s probably not an accident.

Reply:
between including mkldnn_base-inl.h and forward declaration, i'll choose the latter. NDArray is such a basic class, its header file is included in almost every .cc files and many .h files, including mkldnn_base-inl.h. The other option is to define MKLDNNMemory in NDArray, if it's a preferred way. it's a little weird to me.